### PR TITLE
*: Refactor to not mutate global objects

### PR DESCRIPTION
**Pull request by brancz**
Same kind of refactoring as https://github.com/thanos-io/kube-thanos/pull/90.

@squat @bwplotka @metalmatze @kakkoyun @krasi-georgiev

### CODE REVIEW

**Comment by metalmatze** [comment]
Looking really good. :)
The ThanosReceiveController might be a good candidate to experiment with using plain Jsonnet objects as base line. Like writing the ServiceAccount without ksonnet and other components. We could do that in a follow up though, if we would want to try that.

**Comment by brancz** [comment]
Agreed :100: .

File: jsonnet/lib/thanos-receive-controller.libsonnet [hunk 1]
```diff
@@ -1,125 +1,165 @@
...
+    namespace: error 'must provide namespace',
+    version: error 'must provide version',
+    image: error 'must provide image',
+    replicas: error 'must provide replicas',
+    hashrings: error 'must provide hashring configuration',
+
+    commonLabels:: {
+      'app.kubernetes.io/name': 'thanos',
+      'app.kubernetes.io/instance': trc.config.name,
+      'app.kubernetes.io/version': 'v' + trc.config.version,
```
Comment (squat) [inline jsonnet/lib/thanos-receive-controller.libsonnet]: This is probably only true for semver tags and only when those tags don't already start with a 'v'. I'd say let's not prepend anything to the version given by the user. "We expect you to provide the correct version" is easier to say than "We expect you to provide the correct version of your app unless that version starts with a 'v', in which case elide the first character, or if your version is not semver, in which case... x"

**Comment by squat** [comment]
Just one comment, otherwise looks great to me :)

File: jsonnet/lib/thanos-receive-controller.libsonnet (hunk 1 above)
Comment (brancz) [reply jsonnet/lib/thanos-receive-controller.libsonnet]: fixed.

**Pull request merged**